### ISSUE TITLE
How Pipeline::num_threads work and how to leverage for custom operators?

### ISSUE THREAD

**Issue by thiagocrepaldi** [NVIDIA/DALI#1309]
Hello, currently I am evaluating whether DALI can be used to replace Pytorch DataLoader with custom Datasets. In this scenario, we are using several CPUs for each GPU, in a way that multiple Pytorch DataLoader workers (processes) read files on CPU, do some CPU transformation and finally send it to GPU. We can tweak this setup further to make support several GPUs (locally or remotely) with several slave CPUs each.

Looking DALI API, I see there is a num_threads input for the Pipeline and I wonder how it works and what kind of parallelism is achieved. Where are those threads used? For the pipeline itself or by operators? Is it C++ or Python threads? 

Any suggestions on how to achieve N CPUs per GPU data processing using DALI?

I am willing to provide several contributions to implement this idea and others if DALI is a good fit

**Comment by awolant** [comment]
Hi, thanks for the question.

We are happy to hear that you are interested in DALI and are willing to contribute to it. 

Threads controlled with `num_threads` are CPU worker threads that handle data processing. They are C++ threads.
DALI uses more threads to control execution, schedule work on the GPU etc.

@mzient Please tell more and share your perspective on the N CPUs per GPU idea.

**Comment by mzient** [comment]
Hello,
I'll try to explain threading in DALI best as I can:
In general, DALI is a multithreaded library and there are several management threads and worker threads that perform actual processing for CPU operators. The parameter `num_threads` controls the latter.
There's a per-pipeline thread pool object, containing `num_threads` threads, which are used by operators to run processing on. Typically, each sample is scheduled as a separate job and the thread pool picks up queued jobs in submission order and processes them when workers are free.
There's also some work being done on managing threads - specifically, things like data validation, allocation, shape inference and scheduling work for the runners happens in a separate thread. GPU operators use yet another cpu thread.
So, specifying `num_threads = N` does not mean that DALI will use only N threads per pipeline, but these N threads will be the busiest ones.
Now, going back to your question of N CPUs per GPU: DALI pipeline is assigned to a particular GPU - so specifying num_threads = N effecively gives you N cpus per GPU - there are even some affinity settings, but you'd need to experiment to see what works best for you. You can try to tinker with it - specify `set_affinity=True` on your pipeline and provide an environment variable `DALI_AFFINITY_MASK`. Unfortunately, there's no way to set affinity to a pipeline object. Adding explicity affinity mask to pipeline object and propagating it to ThreadPool should be possible to add without ruining too much code.

Thanks,
Michal

P. S. Link to the paragraph about threading in our docs:
https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/advanced_topics.html#thread-affinity

**Comment by thiagocrepaldi** [comment]
Thank you for taking the time to answer my questions, @mzient ! I really appreciate it!

I am still evaluating DALI and based on your answers and some reading, I will enumerate some more questions so that we refer to them more easily.

1) Based on your explanation regarding the worker threads used by operators, my understanding is that if I implement a new operator, I could use those worker threads to achieve parallelism, right? (as opposed to the custom operator create/manage their own threads)

2) Is there a minimal operator example which uses `num_threads` so that I can learn how a *custom* operator could use DALI thread pool? I noticed that most of the operators are focused on image/video, but I am considering to add more operators to handle text, speech files and other formats not currently supported by DALI.

3) Because DALI uses threading for Pipeline/Operator parallelism, all operators that do data transformation/augmentation should be implemented in C++ only, right?

4) Going through examples, I have the impression that no unnecessary data copies happen between `ExternalSource`, `DALIClassificationIterator` or `Pipeline`. In other words, once data is read by a Python `Iterator` and fed to `ExternalSouce` through `iter_setup` inside `Pipeline`, references to data are passed along instead of being copied on each layer, right? I am not considering expected copies between CPU and GPU when the Pipeline Operators specify different devices.

5) Lastly, would DALI be open to accept contributions that add support to different devices other than NVIDIA's, such as specialized TPU? That is an odd question to ask and we don't have any roadmap on that, but I am evaluating hypothetical scenarios on which heavy dependency on DALI is taken and how that would affect users if new hardware technologies emerge.

**Comment by JanuszL** [comment]
> Based on your explanation regarding the worker threads used by operators, my understanding is that if I implement a new operator, I could use those worker threads to achieve parallelism, right? (as opposed to the custom operator create/manage their threads)

You can follow two paths. You can let the executor run in the thread pool operator for each sample in parallel or manage the thread pool on your own. [This](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/operator.h#L251) allows you to manage thread pool on your own, like it is done in [ExternalSource operator](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/util/external_source.cc#L48). Or you can [relay on the executor](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/operator.h#L278) and do like this [simple cast op](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/util/cast.cc#L21) which just process one sample at the time, and the executor does the parallelism.

> Is there a minimal operator example that uses num_threads so that I can learn how a custom operator could use the DALI thread pool? I noticed that most of the operators are focused on image/video, but I am considering to add more operators to handle text, speech files and other formats not currently supported by DALI.

Mentioned [ExternalSource operator](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/util/external_source.cc#L48) manages the thread pool on its own. Keep in mind that its functionality is not bound to `num_threads`, it just uses a thread pool and it is up to the user to set `num_threads` according to HW configuration available.

> Because DALI uses threading for Pipeline/Operator parallelism, all operators that do data transformation/augmentation should be implemented in C++ only, right?

Yes, parallelism can be only fully utilized in C++ code, but we also support [python based operators](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/python_operator/python_operator.html) (mostly for prototyping, not for the performance).

> Going through examples, I have the impression that no unnecessary data copies happen between ExternalSource, DALIClassificationIterator or Pipeline. In other words, once data is read by a Python Iterator and fed to ExternalSouce through iter_setup inside Pipeline, references to data are passed along instead of being copied on each layer, right? I am not considering expected copies between CPU and GPU when the Pipeline Operators specify different devices.

Partially. There buffers between each layer as DALI operators don't work in place. So there are memory transfers, but they are part of the natural processing. So not plain memcopy is made, but operators need separate buffers for input and the output. The good side is that each operator's output could be consumed by any number of operators.

> Lastly, would DALI be open to accepting contributions that add support to different devices other than NVIDIA's, such as specialized TPU? That is an odd question to ask and we don't have any roadmap on that, but I am evaluating hypothetical scenarios on which heavy dependency on DALI is taken and how that would affect users if new hardware technologies emerge.

We are open to any contribution. The main requirement is testability. Now we rely on the internal CI where we probably won't have easy access to specialized TPU, still, we want to make sure that any release we provide doesn't break code we ship. I can imagine that is could be provided as a plugin user could download on its own and we don't officially maintain it (as Nvidia) or the code owner would provide some CI to test against this custom TPU.
And one more down to earth problem which needs to be faced first is that DALI realities on the fact that CUDA run time is present with GPU (even for pure CPU based pipelines). But it should be done to lift that constraint.

**Comment by thiagocrepaldi** [comment]
Thank you, that helps a lot!

> You can follow two paths. You can let the executor run in the thread pool operator for each sample in parallel or manage the thread pool on your own. [This](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/operator.h#L251) allows you to manage thread pool on your own, like it is done in [ExternalSource operator](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/util/external_source.cc#L48). Or you can [relay on the executor](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/operator.h#L278) and do like this [simple cast op](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/util/cast.cc#L21) which just process one sample at the time, and the executor does the parallelism.
> 
> > Is there a minimal operator example that uses num_threads so that I can learn how a custom operator could use the DALI thread pool? I noticed that most of the operators are focused on image/video, but I am considering to add more operators to handle text, speech files and other formats not currently supported by DALI.
> 
> Mentioned [ExternalSource operator](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/util/external_source.cc#L48) manages the thread pool on its own. Keep in mind that its functionality is not bound to `num_threads`, it just uses a thread pool and it is up to the user to set `num_threads` according to HW configuration available.
> 

Regarding
> Keep in mind that its functionality is not bound to `num_threads`, it just uses a thread pool and it is up to the user to set `num_threads` according to HW configuration available.
Does that mean that if I implement `RunImpl(HostWorkspace &ws)`, the input parameter `Pipeline::num_threads` will be ignored and it is the Operator implementer responsibility to limit the number of times `thread_pool.DoWorkWithID()` is called? I was under the impression that the `Pipeline` creates a thread pool for CPU, another for GPU and its `RumImpl` for both `HostWorkspace` and `SampleWorkspace` would be bound to it.

**Comment by JanuszL** [comment]
> Does that mean that a new Operator that implements RunImpl(HostWorkspace &ws) will ignore the input parameter Pipeline::num_threads and it is up to the Operator implementer to limit the number of times thread_pool.DoWorkWithID() is called? I was under the impression that the Pipeline creates a thread pool for CPU, another for GPU and the RumImpl implementations for both HostWorkspace and SampleWorkspace would be bound to num_threads.

I mean that `Pipeline::num_threads` is transparent for the operator implementation, and it shouldn't rely on any particular value of that.

> I was under the impression that the Pipeline creates a thread pool for CPU, another for GPU and the RumImpl implementations for both HostWorkspace and SampleWorkspace would be bound to num_threads.

The thread pool is created only for the CPU operators. GPU one doesn't not any GPU monogramming model implies threading by the definition.

**Comment by thiagocrepaldi** [comment]
Thanks, that helps a lot. Do you have any design/architecture document that could help us asses the complexity in contributing back some features for our use cases?

**Comment by JanuszL** [comment]
Hi,
We don't have anything up to date beyond the GTC presentations https://github.com/NVIDIA/DALI#additional-resources.
Lots of things are under heavy rework the code is the best documentation now.

**Comment by thiagocrepaldi** [comment]
Alright, thank you again!
